### ISSUE TITLE
Catalog price rule applied configurable products not show oldPrice in product list

### ISSUE THREAD

**Issue by cizgidekiadam** [magento/magento2#7624]
### Preconditions

1. Magento 2.1.2
2. PHP 7.0.11

### Steps to reproduce

1. Create a price rule and apply it (for ex. 15% discount)

### Expected result

1. Listing working fine with simple products
<img width="886" alt="screenshot at nov 30 15-26-47" src="https://cloud.githubusercontent.com/assets/4427628/20752384/d7b222fa-b709-11e6-94aa-842c515ba236.png">
2. Same as simples, show both final and old price for configurable products

### Actual result

1. Old price not shown for configurable products
<img width="898" alt="screenshot at nov 30 15-29-44" src="https://cloud.githubusercontent.com/assets/4427628/20752415/fb956c68-b709-11e6-8973-875e6cc60533.png">

**Comment by olysenko** [comment]
Hi,  I cannot reproduce this issue. Please add more details to your description of the steps you followed when identifying this issue.

**Comment by cizgidekiadam** [comment]
@olysenko It just classical usage of catalog price rules.

- Create a rule based on attribute set
- Apply it.

<img width="1065" alt="rule_attribute_set" src="https://cloud.githubusercontent.com/assets/4427628/21041534/488dac88-bdf4-11e6-8792-54e939d79b1e.png">

Actually rule is working on all type of products. 
For simples, discounted prices shown on all areas (listing, search, view etc..)
But configurables, just shown on product view page.

It may be about with price renderer in $block->getProductPrice(). Because in listing, for configurables, magento gets lowest child's price. 

I'm investigating this.

**Comment by cizgidekiadam** [comment]
In addition it may be related with [#6080](https://github.com/magento/magento2/issues/6080)
Our configurables prices are 0. Catalog price calculating by totals of selected attributes (child prices).

**Comment by cizgidekiadam** [comment]
Additional Info: 

After some investigation, i think find a clue
Clean install 2.1.2 with sample-data via composer.

- Create a catalog rule which looks like our original one and apply it. (Attribute set based)
<img width="1286" alt="scr1" src="https://cloud.githubusercontent.com/assets/4427628/21078900/d7c7af1c-bf88-11e6-9a26-01092b36128f.png">
- It is working well with same price children  👍 
<img width="1025" alt="scr2" src="https://cloud.githubusercontent.com/assets/4427628/21078902/0f7a24ee-bf89-11e6-9048-b702d608d4df.png">
<img width="890" alt="scr3" src="https://cloud.githubusercontent.com/assets/4427628/21078931/653a79d2-bf8a-11e6-9ddd-46a5956ab5c5.png">


- But if change price of any child
<img width="888" alt="scr4" src="https://cloud.githubusercontent.com/assets/4427628/21078936/83e0c030-bf8a-11e6-8dc2-106d210c3025.png">


- It shows lowest price as expected but regular price gone same as our situation
<img width="992" alt="scr5" src="https://cloud.githubusercontent.com/assets/4427628/21078917/63711e7c-bf89-11e6-8301-21052058e0a0.png">


- But in product view discount shown. So catalog rule applied all children, not only parent (as expected).
<img width="822" alt="scr6" src="https://cloud.githubusercontent.com/assets/4427628/21078926/18f3c60a-bf8a-11e6-8ea9-54052f3c8923.png">


I hope it will help to solve this.

**Comment by olysenko** [comment]
Hi, after your las comment update I was able to reproduce  issue on clear Magento installation with sample data. Your price index is invalid after price was changed. In case your have setting for indexers UPDATE BY SCHEDULE in menu System->Tools->Index Management. You have to setup cron with indexers reindex job. Or use UPDATE ON SAVE mode for indexers. In this case it will reindex price on product save. Please tell me if my comment was helpful for you.

**Comment by cizgidekiadam** [comment]
Hi @olysenko,
All indexes are update by schedule and crond working fine. Also check database while running reindex. I have verified the calculation is correct and products got discounted price. 

This issue just about showing oldPrice in frontend. Not about rule applying or other.
Product special price did not set for minimal price (lowest price of child products) so final_price.html don't know how about showing price ($block->hasSpecialPrice() in final_price.phtml returns false). 

I did some trick to override it (Magento_Catalog/templates/product/price/final_price.phtml) : 
line around 20 <pre>if($block->getSaleableItem()->getTypeId()==\Magento\ConfigurableProduct\Model\Product\Type\Configurable::TYPE_CODE)
{
    /** ex: \Magento\Catalog\Pricing\Price\ConfiguredPrice */
    /** @var \Magento\Framework\Pricing\Price\PriceInterface $priceModel */
    $priceModel = $block->getPriceType('configured_price');
}</pre>
and line around 32 <pre>if($finalPriceModel->getAmount()->getValue()<$priceModel->getAmount()->getValue())
           $block->getSaleableItem()->setSpecialPrice($finalPriceModel->getAmount()->getValue());
 if ($block->hasSpecialPrice()): ?>

**Comment by olysenko** [comment]
Hi @rbostan, so we need to synchronize our steps. My steps with no bug you can find below:
1. Clean install 2.1.2 with sample-data
2. Set all Indexes mode to UPDATE BY SCHEDULE in menu System->Tools->Index Management
3. Add task  to crontab to execute every 1 minute
* * * * * /usr/bin/php {magento2ce}/bin/magento indexer:reindex
4. Create a catalog rule which looks like our original one and apply it. (Attribute set Top). Discount 20% from original price
5. Go to  front-end. Open category  Women->Tops->Bras & Tanks
All configurable products have two prices actual minimal price and regular price.
For example, product "Breathe-Easy Tank"  two prices available: $27.20 and Regular Price $34.00
6. Change for any simple product "Breathe-Easy Tank"  price to 20$
7. Refresh  category  Women->Tops->Bras & Tanks 
One price is shown $20.00 for product "Breathe-Easy Tank" 
8. Wait 1 minute until indexes will be updated by cron
9. Refresh  category  Women->Tops->Bras & Tanks 
Two prices are shown $16.00 and Regular Price $20.00  for product "Breathe-Easy Tank" 
What is your result after these steps? Do you have another sequence of steps?

**Comment by veloraven** [comment]
According to [contributor guide](http://devdocs.magento.com/guides/v2.1/contributor-guide/contributing.html), tickets without response for two weeks should be closed.
If this issue still reproducible please feel free to create the new one: format new issue according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines): with steps to reproduce, actual result and expected result and specify Magento version.

**Comment by onepack** [comment]
I have the exact same issue on a project that has migrated data from a magento 1 store.
The products are assigned later to the configurable product not on the go while creating the configurable product.
When I apply a special price on a simple product or apply a catalog price rule the old price just never shows just the new discounted price and so it is not clear for a customer that this is a special/ discounted price.
This happens in the grids and on a product detail page.
Only when I click a color option in the detail page the display of the price box gets updated with the text "regular: xx,xx" under the new "special price".

**Comment by onepack** [comment]
This is on a magento 2.1.6 install on a development environment and a production environment with PHP 7.

**Comment by onepack** [comment]
@veloraven this is weird.
I noticed it in more clear reproducible bug-cases. Some moderators seem eager to close them when the 2-week-deadline is reached.
It is understandable that tickets cannot be open all the time but when there is a clear bug it seems to me that solving this to make magento better has more importance than this guideline.
I'm now having the exact same issue and have to create a whole new case while this case could already be solved or at least updated by me and maybe others.

**Comment by joebordo** [comment]
@rbostan can you provide the final price file content? I don't know what to replace or override. thanks in advance!

**Comment by joebordo** [comment]
Hi, Line 22 in the the file: vendor\magento\module-configurable-product\view\base\templates\product\price\final_price.phtml

I replaced This: `<?php if (!$block->isProductList() && $block->hasSpecialPrice()): ?>`
With this: `<?php if ($block->hasSpecialPrice()): ?>`

From onepack, if you want the old price to show in product page (when the swatches are not selected)
Remove the css styling of the "old-price" line 32 (remove everything and leave the old-price class and now it will show on the product page.

hope it helps. someone

**Comment by gediminaskv** [comment]
@joebordo overwrite provided by @rbostan doesn't work on 2.1.7. 

I overwrote like this:

```
<?php
/** @var \Magento\Catalog\Pricing\Render\FinalPriceBox $block */

$productId = $block->getSaleableItem()->getId();

if ($block->getSaleableItem()->getTypeId()==\Magento\ConfigurableProduct\Model\Product\Type\Configurable::TYPE_CODE) {

/** ex: \Magento\Catalog\Pricing\Price\ConfiguredPrice /

/* @var \Magento\Framework\Pricing\Price\PriceInterface $priceModel */

$priceModel = $block->getPriceType('configured_price');

} else {

/** ex: \Magento\Catalog\Pricing\Price\RegularPrice */
/** @var \Magento\Framework\Pricing\Price\PriceInterface $priceModel */
$priceModel = $block->getPriceType('regular_price');
}


/** ex: \Magento\Catalog\Pricing\Price\FinalPrice */
/** @var \Magento\Framework\Pricing\Price\PriceInterface $finalPriceModel */
$finalPriceModel = $block->getPriceType('final_price');
$idSuffix = $block->getIdSuffix() ? $block->getIdSuffix() : '';
$schema = ($block->getZone() == 'item_view') ? true : false;
?>
<?php 
if ($finalPriceModel->getAmount()->getValue() < $priceModel->getAmount()->getValue())
	
$block->getSaleableItem()->setSpecialPrice($finalPriceModel->getAmount()->getValue());

if ($block->hasSpecialPrice()): ?>
..............
..............
.............
```

**Comment by drpit** [comment]
@gediminaskv  It worked form me on 2.1.7 
@joebordo I want to copy this file and put it in my custom theme. What path should I use?

**Comment by saulfm08** [comment]
@joebordo Thanks it works for me!

Someone know how can I change "was" for "old price" ?

**Comment by studioworx-meniconi** [comment]
Sorry to reopen and drag this but im having the same issue on a magento 2.2 with migration data. ive tried the method mentioned by @joebordo but doesnt seem to work as no old price is showing. 

Any advice would be much appreciated.

**Comment by snackbar65** [comment]
After upgrading from Magento 2.1 to  2.2.3 the solution of @joebordo stopped working. Also, when a child product is selected on the product page, the old price is not showing anymore.

**Comment by juandosil** [comment]
I have also tried the method mentioned by @joebordo with Magento 2.2.4 but doesnt seem to work as no old price is showing.

Any advice would be much appreciated.

**Comment by brampta** [comment]
> Hi, Line 22 in the the file: vendor\magento\module-configurable-product\view\base\templates\product\price\final_price.phtml
> 
> I replaced This: `<?php if (!$block->isProductList() && $block->hasSpecialPrice()): ?>`
> With this: `<?php if ($block->hasSpecialPrice()): ?>`
> 
> From onepack, if you want the old price to show in product page (when the swatches are not selected)
> Remove the css styling of the "old-price" line 32 (remove everything and leave the old-price class and now it will show on the product page.
> 
> hope it helps. someone

I found that this solution works for me on 2.2.6, now old prices are showing for configurable products in the catalog, hower this causes a new bug, on product pages the old price is initially shown for a second or 2 and then disappears in a flash :(. When selecting an option it goes back to displaying but that was always like that. Seems like there is some Ajax rewriting the price after the fact on the product page.

======edit:
found a fix for the bug created by the joebordo fix: https://github.com/magento/magento2/issues/8670

**Comment by deanpeterson** [comment]
> > Hi, Line 22 in the the file: vendor\magento\module-configurable-product\view\base\templates\product\price\final_price.phtml
> > I replaced This: `<?php if (!$block->isProductList() && $block->hasSpecialPrice()): ?>`
> > With this: `<?php if ($block->hasSpecialPrice()): ?>`
> > From onepack, if you want the old price to show in product page (when the swatches are not selected)
> > Remove the css styling of the "old-price" line 32 (remove everything and leave the old-price class and now it will show on the product page.
> > hope it helps. someone
> 
> I found that this solution works for me on 2.2.6, now old prices are showing for configurable products in the catalog, hower this causes a new bug, on product pages the old price is initially shown for a second or 2 and then disappears in a flash :(. When selecting an option it goes back to displaying but that was always like that. Seems like there is some Ajax rewriting the price after the fact on the product page.
> 
> ======edit:
> found a fix for the bug created by the joebordo fix: #8670

@brampta What fix did you find for the bug created by the joebordo fix? I'm on 2.2.3 and I get the old price to flash for a second and then it is replaced with the sale price. So the sale price is displayed and the Original Price, also displays the sale price.

**Comment by gitmoorka** [comment]
I am on 2.3.4 and both prices show up after the fix described above. However, I only have special price on **some** variants of the configurable product, and the old prices show up for all variations. As a result, I have Old price 99EUR new price 99EUR which looks like I am trying to fool my customer. Is there a fix for that?
